### ISSUE TITLE
Add Support for Cupy, Dask and JAX arrays

### ISSUE THREAD

**Issue by matthieumeo** [matthieumeo/pyxu#7]
## Extend scope of functions/methods to Cupy, Dask and JAX arrays (Duck arrays)

- [x] Add decorator `@infer_array_module` to all numpy-backed functions of the package to extend their scope to *Cupy*, *JAX* and *Dask* arrays and hence provide support for GPU/TPU and out-of-core/distributed computations. 
The decorator should infer the array module to be used (`numpy`, `cupy` or `dask`) from the input duck array fed to the function (similarly to `cp.get_array_module()`). The code should be made module-agnostic by replacing all `np.function` by `_xp.function` where `_xp` is dynamically inferred at run-time by the decorator. 

Example:

```python
import numpy as np
import cupy as cp
import dask.array as da
import jax.numpy as jnp
import types
from typing import Callable, Optional

def infer_array_module(decorated_object_type='method'):
  def make_decorator(call_fun: Callable):
    def wrapper(*args, **kwargs):
      if decorated_object_type == 'method':
        arr = args[1] #First argument is self
      else:
        arr = args[0]

      if isinstance(arr, cp.ndarray):
        xp = cp
      elif isinstance(arr, da.core.Array):
        xp = da 
      elif isinstance(arr, jnp.ndarray):
        xp = jnp
      else:
        xp = np # Fall back to Numpy backend if unknown array type
      kwargs['_xp'] = xp
      return call_fun(*args, **kwargs)
    return wrapper
  return make_decorator

class FFTOp(object):

  def __init__(self):
    pass

  @infer_array_module()
  def __call__(self, x, _xp: Optional[types.ModuleType] = None):
    return _xp.fft.fft(x)
```

## Cupy specifics:

- [x] Control better `dtypes` of `Map`, `LinearOperator` and `Functional` objects as well as algorithms (`float64` computations on GPUs are notoriously slower than `float32` computations)
- [x] Cupy should remain an optional dependency. Any call to the `cupy` module should hence be conditional to this dependency being installed.
- [ ] Benchmark cupy vs numpy computations.

## Dask specifics:

- [x] Output of callables are  by default lazy dask arrays. In iterative algorithms, lazy arrays should be computed via the method `persist()`  (equivalent to `compute()` on a single machine but keeps the computation result distributed when working with multiple nodes). This is necessary since the stopping criterion to most algorithms cannot be computed lazily. Moreover, computing the dask arrays avoid overly long dask computation graphs which comes with much overhead. 

## JAX specifics: 
- [x] JAX arrays are immutable so inplace computations in the code must be updated (potentially cumbersome).

**Comment by andresmntr** [comment]
# Compatibility checklist for pycsou/pycsou/


#### ./core/map.py
- [x] Map
- [x] MapShifted
- [x] MapSum
- [x] MapComp
- [x] DifferentiableMap
- [x] DiffMapShifted
- [x] DiffMapSum
- [x] DiffMapComp
- [x] MapStack
- [x] MapVStack
- [x] MapHStack
- [x] DiffMapStack
- [x] DiffMapVStack
- [x] DiffMapHStack

#### ./core/linop.py
- [x] LinearOperator
- [x] AdjointLinearOperator
- [x] TransposeLinearOperator
- [x] LinOpSum
- [x] LinOpComp
- [x] SymmetricLinearOperator
- [x] UnitaryOperator
- [x] LinOpPinv

#### ./core/functional.py
- [x] Functional
- [x] DifferentiableFunctional
- [x] LinearFunctional
- [x] ProximableFunctional
- [x] ProcFuncPostComp
- [x] ProxFuncAffineSum
- [x] ProxFuncPreComp
- [x] ProxFuncCompUnitOp



#### ./linop/base.py

- [x] PyLopLinearOperator
- [x] ExplicitLinearOperator
- [x] DenseLinearOperator
- [x] SparseLinearOperator
- [x] DaskLinearOperator
- [x] LinOpStack
- [x] LinOpVStack
- [x] LinOpHStack
- [x] BlockOperator
- [x] BlockDiagonalOperator
- [x] DiagonalOperator
- [x] IdentityOperator
- [x] NullOperator
- [x] HomothetyMap
- [ ] PolynomialLinearOperator
- [x] KroneckerProduct
- [ ] KroneckerSum
- [x] KhatriRaoProduct

#### ./linop/conv.py

- [ ] Convolve1D
- [ ] Convolve2D
- [ ] MovingAverage1D
- [ ] MovingAverage2D

#### ./linop/diff.py

- [ ] FirstDerivative
- [ ] SecondDerivative
- [ ] GeneralisedDerivative
- [ ] FirstDerivative
- [ ] SecondDerivative
- [ ] DirectionalGradient
- [ ] DirectionalLaplacian
- [ ] Gradient
- [ ] Laplacian
- [ ] GeneralisedLaplacian
- [ ] Integration1D

#### ./linop/sampling.py

- [x] SubSampling
- [ ] Masking
- [x] DownSampling
- [ ] Pooling
- [ ] NNSampling
- [ ] GeneralisedVandermonde
- [ ] MappedDistanceMatrix
